src/network/networkd-link.c

```diff
@@ -2638,11 +2681,23 @@ int link_update(Link *link, sd_netlink_message *m) {
                 r = link_add(manager, m, &link);
                 if (r < 0)
                         return r;
+
+                hashmap_remove_value(link->manager->links_by_name, link->ifname, link);
+                hashmap_put(link->manager->links_by_name, ifname, link);
         }
 
         r = sd_netlink_message_read_strv(m, IFLA_PROP_LIST, IFLA_ALT_IFNAME, &s);
-        if (r >= 0)
+        if (r >= 0) {
+                char **n;
+
+                STRV_FOREACH(n, link->alternative_names)
+                        hashmap_remove_value(link->manager->links_by_name, *n, link);
+
+                STRV_FOREACH(n, s)
+                        hashmap_put(link->manager->links_by_name, *n, link);
```

Comment:
oom not checked.

i guess the negative effect of this is not too bad, hence might make sense, to just log about this and proceed, i.e. `log_warning_errno()`